mart/attack/adversary.py

```diff
@@ -100,16 +100,17 @@ def __init__(
         """
         super().__init__()
 
-        self.perturber = perturber
+        # So that perturbation is not in module.parameters()
+        self.perturber = [perturber]
```

Comment:
Why does this matter? That is, for what piece of code does this affect?

Reply:
Okay. I think I see the issue. Now that you're using `mart.attack.Perturber` directly, that exposes parameters whereas previously `mart.attack.BatchPerturber` did not.

There has to be a better way of doing this...

Reply:
I think what you want to do is override `def parameters` and filter parameters by name.

Reply:
However, it is also unclear to me how this affects the placement of the perturber's parameters. How do they end up on the GPU?

Reply:
I don't see an easy path for parameter filtering, unless you want to change `_modules` or `_parameters`.

The recursion happens in `torch.nn.Module._modules` to get all sub modules,
https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/module.py#L1889

then they use `Module._parameters` to get parameters of each module.
https://github.com/pytorch/pytorch/blob/v1.13.1/torch/nn/modules/module.py#L1735

Reply:
Why can't you `filter(lambda name, _: name.contains("perturber"), super().parameters())`? or something like that?

Reply:
Because `parameters()` of `Adversary` or `Perturber` is not called when enumerating parameters of a model.

Reply:
Okay, so now I don't understand what `self.perturber = [perturber]` is solving then? My assumption was there is some code that is calling `IterativeGenerator.parameters` and that was returning the parameters of the `perturber` because recursion. What am I missing?

Reply:
It solves two issues:
1. Experiments try to count the number of params of a model before training, and the uninitialized parameter in Perturber leads to an exception. `self.perturber = [perturber]` make it ignore parameters in a perturber.
2. Experiments try to save weights after training, and `self.perturber = [perturber]` makes sure that `perturbation` is not saved, so that we can later load the weights in the strict mode.

No, there's no one calling `IterativeGenerator.parameters`. The parameters are recursively searched with `torch.nn.Module._modules`. `self.perturber = [perturber]` excludes a perturber from `torch.nn.Module._modules`, which is why it works.

Reply:
Okay this is good.

For (1), how does counting the number of parameters work? There must be some API that codes uses to figure out that perturbed is an nn.Parameter?

For (2), that can be solved by overriding state_dict or setting persistent=False on the buffer.

Reply:
@mzweilin: I think I made this work?